From 0b45d1d36e9d0d3c033d8de012b2652ffa892568 Mon Sep 17 00:00:00 2001
From: linmin <linmin@eswincomputing.com>
Date: Thu, 21 Nov 2024 08:41:41 +0800
Subject: [PATCH 291/416] WIN2030-14758:refactor: refact split dmabuf

Changelogs:
1.Generate a new splitted sgt from the parent's dmabuf, and export the splitted sgt
  as a new dmabuf.

Signed-off-by: linmin <linmin@eswincomputing.com>
Change-Id: I0dbdf1feae39f7824466d0741deb0092b8afb005
---
 drivers/memory/eswin/es_mmz_vb/mmz_vb.c       |  23 +-
 .../dmabuf-heap-import-helper.c               | 635 ++++++++++++++----
 include/linux/dmabuf-heap-import-helper.h     |  40 +-
 3 files changed, 552 insertions(+), 146 deletions(-)

diff --git a/drivers/memory/eswin/es_mmz_vb/mmz_vb.c b/drivers/memory/eswin/es_mmz_vb/mmz_vb.c
index 8fad791695d7..9876cc2fe319 100644
--- a/drivers/memory/eswin/es_mmz_vb/mmz_vb.c
+++ b/drivers/memory/eswin/es_mmz_vb/mmz_vb.c
@@ -1,11 +1,22 @@
-// SPDX-License-Identifier: GPL-2.0-only
+// SPDX-License-Identifier: GPL-2.0
 /*
  * ESWIN MMZ VB driver. MMZ VB stands for Media Memory Zone Video Buffer.
  *
- * Copyright 2024 Beijing ESWIN Computing Technology Co., Ltd.
- *   Authors:
- *    LinMin<linmin@eswincomputing.com>
+ * Copyright 2024, Beijing ESWIN Computing Technology Co., Ltd.. All rights reserved.
  *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Authors: Min Lin <linmin@eswincomputing.com>
  */
 
 #include <linux/dma-buf.h>
@@ -2076,7 +2087,7 @@ static int vb_blk_to_pool(struct esVB_BLOCK_TO_POOL_CMD_S *blkToPoolCmd)
 	int ret = 0;
 	struct dma_buf *dmabuf;
 	struct mmz_vb_buffer *buffer;
-	struct esw_export_buffer_info *splittedBuffer;
+	struct eswin_split_buffer *splittedBuffer;
 	struct dma_buf *blkDmabuf;
 	bool isSplittedBlk;
 
@@ -2119,7 +2130,7 @@ static int vb_get_blk_offset(struct esVB_GET_BLOCKOFFSET_CMD_S *getBlkOffsetCmd)
 	int ret = 0;
 	struct dma_buf *dmabuf;
 	struct mmz_vb_buffer *buffer;
-	struct esw_export_buffer_info *splittedBuffer;
+	struct eswin_split_buffer *splittedBuffer;
 	struct dma_buf *blkDmabuf;
 	__u64 blkSize, offsetInPool;
 
diff --git a/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c b/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
index 8e1e14487a31..2a2877f848b2 100644
--- a/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
+++ b/drivers/memory/eswin/es_rsvmem_heap/dmabuf-heap-import-helper.c
@@ -1,21 +1,37 @@
-// SPDX-License-Identifier: GPL-2.0-only
+// SPDX-License-Identifier: GPL-2.0
 /*
  * ESWIN DMABUF heap helper APIs
  *
- * Copyright 2024 Beijing ESWIN Computing Technology Co., Ltd.
- *   Authors:
- *    LinMin<linmin@eswincomputing.com>
+ * Copyright 2024, Beijing ESWIN Computing Technology Co., Ltd.. All rights reserved.
  *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Authors: Min Lin <linmin@eswincomputing.com>
  */
 
 #include <linux/export.h>
 #include <linux/slab.h>
 #include <linux/errno.h>
+#include <linux/highmem.h>
 #include <linux/fcntl.h>
 #include <linux/scatterlist.h>
+#include <linux/miscdevice.h>
+#include <linux/dma-map-ops.h>
 #include <linux/dma-heap.h>
 #include <linux/dmabuf-heap-import-helper.h>
 
+static struct device *split_dmabuf_dev;
+
 struct drm_prime_member {
 	struct dma_buf *dma_buf;
 	uint64_t handle;
@@ -569,220 +585,571 @@ void common_dmabuf_heap_rsv_iova_uninit(struct heap_root *root)
 }
 EXPORT_SYMBOL(common_dmabuf_heap_rsv_iova_uninit);
 
-struct esw_exp_attachment {
-	struct heap_mem *hmem;
-	struct sg_table table;
+
+struct eswin_split_attachment {
 	struct device *dev;
-	struct heap_root root;
+	struct sg_table *table;
+	struct list_head list;
+	bool mapped;
 };
 
+static struct sg_table *dup_sg_table(struct sg_table *table)
+{
+	struct sg_table *new_table;
+	int ret, i;
+	struct scatterlist *sg, *new_sg;
+
+	new_table = kzalloc(sizeof(*new_table), GFP_KERNEL);
+	if (!new_table)
+		return ERR_PTR(-ENOMEM);
+
+	ret = sg_alloc_table(new_table, table->orig_nents, GFP_KERNEL);
+	if (ret) {
+		kfree(new_table);
+		return ERR_PTR(-ENOMEM);
+	}
+
+	new_sg = new_table->sgl;
+	for_each_sgtable_sg(table, sg, i) {
+		sg_set_page(new_sg, sg_page(sg), sg->length, sg->offset);
+		new_sg = sg_next(new_sg);
+	}
+
+	return new_table;
+}
+
 // #define PRINT_ORIGINAL_SPLITTERS 1
-static int esw_common_heap_attach(struct dma_buf *dmabuf,
+static int eswin_split_heap_attach(struct dma_buf *dmabuf,
 			   struct dma_buf_attachment *attachment)
 {
-	struct esw_export_buffer_info *buffer = dmabuf->priv;
-	struct esw_exp_attachment *a;
-	int out_mapped_nents[1];
-	int ret = 0;
-	struct sg_table *sgt = NULL;
-	struct scatterlist *sg;
-	int i;
-	size_t size, len;
+	struct eswin_split_buffer *buffer = dmabuf->priv;
+	struct eswin_split_attachment *a;
+	struct sg_table *table;
 
 	a = kzalloc(sizeof(*a), GFP_KERNEL);
 	if (!a)
 		return -ENOMEM;
 
-	a->dev = attachment->dev;
-
-	common_dmabuf_heap_import_init(&a->root, a->dev);
-	a->hmem = common_dmabuf_heap_import_from_user(&a->root, buffer->dbuf_fd);
-	if (IS_ERR(a->hmem))
-		return PTR_ERR(a->hmem);
-
-	ret = sg_split(a->hmem->sgt->sgl, a->hmem->sgt->nents, buffer->slice.offset, 1, &buffer->slice.len,
-					&a->table.sgl, &out_mapped_nents[0], GFP_KERNEL);
-	if (ret) {
-		common_dmabuf_heap_release(a->hmem);
+	table = dup_sg_table(&buffer->sg_table);
+	if (IS_ERR(table)) {
 		kfree(a);
-		return ret;
-	}
-	a->table.nents = out_mapped_nents[0];
-	a->table.orig_nents = out_mapped_nents[0];
-	sgt = &a->table;
-	#ifdef PRINT_ORIGINAL_SPLITTERS
-	{
-		pr_info("%s:orig:sgt->orig_nents=%d, out_mapped_nents[0]=%d\n",
-			__func__, sgt->orig_nents, out_mapped_nents[0]);
-		for_each_sg(sgt->sgl, sg, sgt->orig_nents, i) {
-			pr_info("orig[%d]:sg->length=0x%x, sg_dma_len=0x%x, sg_phys=0x%lx\n",
-				i, sg->length, sg_dma_len(sg), (unsigned long)sg_phys(sg));
-		}
-	}
-	#endif
-	/* Re-format the splitted sg list in the actual slice len */
-	{
-		size = buffer->slice.len;
-		for_each_sg(sgt->sgl, sg, sgt->orig_nents, i) {
-			if (sg->length >= size) {
-				sg->length = size;
-				sg_dma_len(sg) = size;
-				sg_mark_end(sg);
-				pr_debug("refmt[%d]:sg->length=0x%x, sg_dma_len=0x%x, sg_phys=0x%lx\n",
-					i, sg->length, sg_dma_len(sg), (unsigned long)sg_phys(sg));
-				break;
-			}
-			len = min_t(size_t, size, sg->length);
-			size -= len;
-			pr_debug("refmt[%d]:sg->length=0x%x, sg_dma_len=0x%x, sg_phys=0x%lx\n",
-				i, sg->length, sg_dma_len(sg), (unsigned long)sg_phys(sg));
-		}
-		sgt->orig_nents = sgt->nents = i + 1;
+		return -ENOMEM;
 	}
 
+	a->table = table;
+	a->dev = attachment->dev;
+	INIT_LIST_HEAD(&a->list);
+	a->mapped = false;
+
 	attachment->priv = a;
 
-	return ret;
+	mutex_lock(&buffer->lock);
+	list_add(&a->list, &buffer->attachments);
+	mutex_unlock(&buffer->lock);
+
+	return 0;
 }
 
-static void esw_common_heap_detach(struct dma_buf *dmabuf,
+static void eswin_split_heap_detach(struct dma_buf *dmabuf,
 			    struct dma_buf_attachment *attachment)
 {
-	struct esw_exp_attachment *a = attachment->priv;
+	struct eswin_split_buffer *buffer = dmabuf->priv;
+	struct eswin_split_attachment *a = attachment->priv;
 
-	kfree(a->table.sgl);
-	common_dmabuf_heap_release(a->hmem);
-	common_dmabuf_heap_import_uninit(&a->root);
+	mutex_lock(&buffer->lock);
+	list_del(&a->list);
+	mutex_unlock(&buffer->lock);
+
+	sg_free_table(a->table);
+	kfree(a->table);
 	kfree(a);
 }
 
-static struct sg_table *esw_common_map_dma_buf(struct dma_buf_attachment *attachment,
+static struct sg_table *eswin_split_map_dma_buf(struct dma_buf_attachment *attachment,
 					     enum dma_data_direction direction)
 {
-	struct esw_exp_attachment *a = attachment->priv;
-	return &a->table;
+	struct eswin_split_attachment *a = attachment->priv;
+	struct sg_table *table =a->table;
+	int ret;
+	unsigned long attrs = DMA_ATTR_SKIP_CPU_SYNC;
+
+	/* Skipt cache sync, since it takes a lot of time when import to device.
+	*  It's the user's responsibility for guaranteeing the cache coherency by
+	   flusing cache explicitly before importing to device.
+	*/
+	ret = dma_map_sgtable(attachment->dev, table, direction, attrs);
+
+	if (ret)
+		return ERR_PTR(-ENOMEM);
+	a->mapped = true;
+	return table;
 }
 
-static void esw_common_unmap_dma_buf(struct dma_buf_attachment *attachment,
+static void eswin_spilt_unmap_dma_buf(struct dma_buf_attachment *attachment,
 				   struct sg_table *table,
 				   enum dma_data_direction direction)
 {
-}
+	struct eswin_split_attachment *a = attachment->priv;
+	unsigned long attrs = DMA_ATTR_SKIP_CPU_SYNC;
 
-static int esw_common_mmap(struct dma_buf *dmabuf, struct vm_area_struct *vma)
-{
-	struct esw_export_buffer_info *buffer = dmabuf->priv;
-	// printk("%s enter\n", __func__);
-	return dma_buf_mmap(buffer->dmabuf, vma, buffer->slice.offset >> PAGE_SHIFT);
+	a->mapped = false;
+
+	/* Skipt cache sync, since it takes a lot of time when unmap from device.
+	*  It's the user's responsibility for guaranteeing the cache coherency after
+	   the device has done processing the data.(For example, CPU do NOT read untill
+	   the device has done)
+	*/
+	dma_unmap_sgtable(attachment->dev, table, direction, attrs);
 }
 
-static int esw_common_dma_buf_begin_cpu_access(struct dma_buf *dmabuf,
+static int eswin_split_dma_buf_begin_cpu_access(struct dma_buf *dmabuf,
 					     enum dma_data_direction direction)
 {
-	struct esw_export_buffer_info *buffer = dmabuf->priv;
-	return dma_buf_begin_cpu_access(buffer->dmabuf, direction);
+	struct eswin_split_buffer *buffer = dmabuf->priv;
+	struct sg_table *table = &buffer->sg_table;
+	struct scatterlist *sg;
+	int i;
+
+	mutex_lock(&buffer->lock);
+
+	if (buffer->vmap_cnt)
+		invalidate_kernel_vmap_range(buffer->vaddr, buffer->len);
+
+	/* Since the cache sync was skipped when eswin_rsvmem_heap_map_dma_buf/eswin_rsvmem_heap_unmap_dma_buf,
+	   So force cache sync here when user call ES_SYS_MemFlushCache, eventhough there
+	   is no device attached to this dmabuf.
+	*/
+	for_each_sg(table->sgl, sg, table->orig_nents, i)
+		arch_sync_dma_for_cpu(sg_phys(sg), sg->length, direction);
+
+	mutex_unlock(&buffer->lock);
+
+	return 0;
 }
 
-static int esw_common_dma_buf_end_cpu_access(struct dma_buf *dmabuf,
+static int eswin_split_dma_buf_end_cpu_access(struct dma_buf *dmabuf,
 					   enum dma_data_direction direction)
 {
-	struct esw_export_buffer_info *buffer = dmabuf->priv;
-	return dma_buf_end_cpu_access(buffer->dmabuf, direction);
+	struct eswin_split_buffer *buffer = dmabuf->priv;
+	struct sg_table *table = &buffer->sg_table;
+	struct scatterlist *sg;
+	int i;
+
+	mutex_lock(&buffer->lock);
+
+	if (buffer->vmap_cnt)
+		flush_kernel_vmap_range(buffer->vaddr, buffer->len);
+
+	/* Since the cache sync was skipped while eswin_rsvmem_heap_map_dma_buf/eswin_rsvmem_heap_unmap_dma_buf,
+	   So force cache sync here when user call ES_SYS_MemFlushCache, eventhough there
+	   is no device attached to this dmabuf.
+	*/
+	for_each_sg(table->sgl, sg, table->orig_nents, i)
+		arch_sync_dma_for_device(sg_phys(sg), sg->length, direction);
+	mutex_unlock(&buffer->lock);
+
+	return 0;
 }
 
-static int esw_common_heap_vmap(struct dma_buf *dmabuf, struct dma_buf_map *map)
+static int eswin_split_mmap(struct dma_buf *dmabuf, struct vm_area_struct *vma)
 {
-	struct esw_export_buffer_info *buffer = dmabuf->priv;
-	struct dma_buf_map pmap;
+	struct eswin_split_buffer *buffer = dmabuf->priv;
+	struct sg_table *table = &buffer->sg_table;
+	unsigned long addr = vma->vm_start;
+	unsigned long pgoff = (table->sgl->offset >> PAGE_SHIFT);
+	// unsigned long pgoff = (buffer->slice.offset >> PAGE_SHIFT);
+	unsigned long size_remaining = vma->vm_end - vma->vm_start;//vma_pages(vma);
+	struct sg_page_iter piter;
 	int ret;
 
-	ret = dma_buf_vmap(buffer->dmabuf, &pmap);
+	/* Mapping secure_memory with cached proprty to user space for CPU is NOT permitted */
+	if (unlikely(!strncmp("secure_memory", dmabuf->exp_name, 13))) {
+		if (!(vma->vm_flags & VM_NORESERVE))
+			return -EPERM;
+	}
+
+	if ((vma->vm_flags & (VM_SHARED | VM_MAYSHARE)) == 0)
+		return -EINVAL;
+
+	/* vm_private_data will be used by eswin-ipc-scpu.c.
+	    ipc will import this dmabuf to get iova.
+	*/
+	vma->vm_private_data = dmabuf;
+
+	/* support mman flag MAP_SHARED_VALIDATE | VM_NORESERVE, used to map uncached memory to user space.
+	   Users should guarantee this buffer has been flushed to cache already.
+	 */
+	if (vma->vm_flags & VM_NORESERVE) {
+		vm_flags_clear(vma, VM_NORESERVE);
+		vma->vm_page_prot = pgprot_dmacoherent(vma->vm_page_prot);
+		/* skip sync cache, users should guarantee the cache is clean after done using it in
+		   cached mode(i.e, ES_SYS_Mmap(SYS_CACHE_MODE_CACHED))
+		*/
+	}
+	pr_debug("%s, vma->vm_start:0x%lx, vma->vm_end:0x%lx\n", __func__, vma->vm_start, vma->vm_end);
+	pr_debug("%s, size_remaining:0x%lx, pgoff:0x%lx, vma->vm_pgoff:0x%lx, dmabuf->size:0x%lx, start_phys:0x%llx, pfn sg_page(sg):0x%lx,sg->length=0x%x\n",
+		__func__, size_remaining, pgoff, vma->vm_pgoff, dmabuf->size, sg_phys(table->sgl),page_to_pfn(sg_page(table->sgl)), table->sgl->length);
+
+	for_each_sgtable_page(table, &piter, pgoff) {
+		struct page *page = sg_page_iter_page(&piter);
+		// pr_debug("addr:0x%lx, page_to_pfn:0x%lx\n", addr, page_to_pfn(page));
+		ret = remap_pfn_range(vma, addr, page_to_pfn(page), PAGE_SIZE,
+				      vma->vm_page_prot);
+		if (ret)
+			return ret;
+		addr += PAGE_SIZE;
+		if (addr >= vma->vm_end)
+			return 0;
+	}
+
+	return 0;
+}
+
+static void *eswin_split_heap_do_vmap(struct dma_buf *dmabuf)
+{
+	struct eswin_split_buffer *buffer = dmabuf->priv;
+	pgprot_t prot = PAGE_KERNEL;
+	struct sg_table *table = &buffer->sg_table;
+	int npages = PAGE_ALIGN(buffer->slice.len) / PAGE_SIZE;
+	struct page **pages = vmalloc(sizeof(struct page *) * npages);
+	struct page **tmp = pages;
+	struct sg_page_iter piter;
+	void *vaddr;
+	unsigned long pgoff = (table->sgl->offset >> PAGE_SHIFT);
+
+	if (!pages)
+		return ERR_PTR(-ENOMEM);
+
+	for_each_sgtable_page(table, &piter, pgoff) {
+		WARN_ON(tmp - pages >= npages);
+		*tmp++ = sg_page_iter_page(&piter);
+	}
+
+	/* The property of this dmabuf in kernel space is determined by heap alloc with fd_flag. */
+	if (buffer->fd_flags & O_DSYNC) {
+		prot = pgprot_dmacoherent(PAGE_KERNEL);
+		pr_debug("%s syport uncached kernel dmabuf!, prot=0x%x\n", __func__, (unsigned int)pgprot_val(prot));
+	}
+	else {
+		pr_debug("%s memport cached kernel dmabuf!\n", __func__);
+	}
+
+	vaddr = vmap(pages, npages, VM_MAP, prot);
+	vfree(pages);
+
+	if (!vaddr)
+		return ERR_PTR(-ENOMEM);
+
+	return vaddr;
+}
+
+static int eswin_split_heap_vmap(struct dma_buf *dmabuf, struct dma_buf_map *map)
+{
+	struct eswin_split_buffer *buffer = dmabuf->priv;
+	void *vaddr;
+	int ret = 0;
 
-	map->is_iomem = false;
-	map->vaddr_iomem = pmap.vaddr_iomem + buffer->slice.offset;
+	mutex_lock(&buffer->lock);
+	if (buffer->vmap_cnt) {
+		buffer->vmap_cnt++;
+		dma_buf_map_set_vaddr(map, buffer->vaddr);
+		goto out;
+	}
+
+	vaddr = eswin_split_heap_do_vmap(dmabuf);
+	if (IS_ERR(vaddr)) {
+		ret = PTR_ERR(vaddr);
+		goto out;
+	}
+	buffer->vaddr = vaddr;
+	buffer->vmap_cnt++;
+	dma_buf_map_set_vaddr(map, buffer->vaddr);
+out:
+	mutex_unlock(&buffer->lock);
 
 	return ret;
 }
 
-static void esw_common_heap_vunmap(struct dma_buf *dmabuf, struct dma_buf_map *map)
+static void eswin_split_heap_vunmap(struct dma_buf *dmabuf, struct dma_buf_map *map)
 {
-	struct esw_export_buffer_info *buffer = dmabuf->priv;
-	struct dma_buf_map pmap = *map;
+	struct eswin_split_buffer *buffer = dmabuf->priv;
 
-	pmap.vaddr_iomem -= buffer->slice.offset;
-	dma_buf_vunmap(buffer->dmabuf, &pmap);
+	mutex_lock(&buffer->lock);
+	if (!--buffer->vmap_cnt) {
+		vunmap(buffer->vaddr);
+		buffer->vaddr = NULL;
+	}
+	mutex_unlock(&buffer->lock);
+	dma_buf_map_clear(map);
 }
 
-static void esw_common_dma_buf_release(struct dma_buf *dmabuf)
+/* split parent dmabuf, and generate a new child sgt */
+static int eswin_get_split_dmabuf(struct eswin_split_buffer *split_buffer)
 {
-	struct esw_export_buffer_info *buffer = dmabuf->priv;
+	struct dma_buf *par_dma_buf = split_buffer->dmabuf;
+	u64 offset = split_buffer->slice.offset;
+	size_t size = split_buffer->slice.len;
+	 struct sg_table *splitted_sgt = &split_buffer->sg_table;
+	 struct sg_table *orig_splitted_sgt = &split_buffer->orig_sg_table;
+	struct dma_buf_attachment *attach;
+	struct dma_heap_attachment *a;
+	struct sg_table *par_origin_table;
+	int out_nents[1];
+	struct scatterlist *sg;
+	struct scatterlist *split_sg;
+	size_t size_remaining, len;
+	int i;
+	int ret = 0;
+
+	if (split_dmabuf_dev == NULL) {
+		pr_err("split_dmabuf_dev has not been created!!!\n");
+		return -EINVAL;
+	}
+
+	get_dma_buf(par_dma_buf);
+	attach = dma_buf_attach(par_dma_buf, split_dmabuf_dev);
+	if (IS_ERR(attach)) {
+		ret = PTR_ERR(attach);
+		dma_buf_put(par_dma_buf);
+		return ret;
+	}
+
+	a = (struct dma_heap_attachment *)attach->priv;
+	par_origin_table = a->table;
+	#ifdef PRINT_ORIGINAL_SPLITTERS
+	{
+		pr_debug("%s:parent[0x%px]:sgt->orig_nents=%d, nents=%d, sg_nents(in) %d\n",
+			__func__, par_dma_buf,
+			par_origin_table->orig_nents, par_origin_table->nents,
+			sg_nents(par_origin_table->sgl));
+		for_each_sg(par_origin_table->sgl, sg, par_origin_table->orig_nents, i) {
+			pr_debug("parent[0x%px]:orig[%d]:sg->offset=0x%x ,sg->length=0x%x, sg_dma_len=0x%x, sg_phys=0x%lx\n",
+				par_dma_buf, i, sg->offset, sg->length, sg_dma_len(sg), (unsigned long)sg_phys(sg));
+		}
+	}
+	#endif
 
-	// printk("%s %d\n", __func__, __LINE__);
+	/* split unmaped/original sgt of parent */
+	ret = sg_split(par_origin_table->sgl, 0, offset, 1, &size,
+					&split_sg, &out_nents[0], GFP_KERNEL);
+	if (ret) {
+		pr_err("Failed to split from parents's sgt\n");
+	}
 
-	dma_buf_put(buffer->dmabuf);
-	kfree(buffer);
+	orig_splitted_sgt->orig_nents = out_nents[0];
+	orig_splitted_sgt->nents = out_nents[0];
+	orig_splitted_sgt->sgl = split_sg;
+
+	splitted_sgt->orig_nents = out_nents[0];
+	splitted_sgt->nents = out_nents[0];
+	/* skip the first sgl if it's length is 0 */
+	if (split_sg->length == 0) {
+		splitted_sgt->sgl = sg_next(split_sg);
+		splitted_sgt->orig_nents--;
+		splitted_sgt->nents--;
+	}
+	else {
+		splitted_sgt->sgl = split_sg;
+	}
+	/* Re-format the splitted sg list in the actual slice len */
+	pr_debug("%s:%d, split[0x%px]:out_nents[0]=%d, Re-format the splitted sgl.\n", __func__, __LINE__, par_dma_buf, out_nents[0]);
+	size_remaining = size;
+
+	for_each_sgtable_sg(splitted_sgt, sg, i) {
+		len = min_t(size_t, size_remaining, sg->length);
+		size_remaining -= len;
+		if (size_remaining == 0) {
+			sg->length = len;
+			sg_mark_end(sg);
+			pr_debug("split[0x%px]:orig[%d]:sg->offset=0x%x ,sg->length=0x%x, sg_dma_len=0x%x, sg_phys=0x%lx\n",
+				par_dma_buf, i, sg->offset, sg->length, sg_dma_len(sg), (unsigned long)sg_phys(sg));
+			break;
+		}
+			pr_debug("split[0x%px]:orig[%d]:sg->offset=0x%x ,sg->length=0x%x, sg_dma_len=0x%x, sg_phys=0x%lx\n",
+				par_dma_buf, i, sg->offset, sg->length, sg_dma_len(sg), (unsigned long)sg_phys(sg));
+	}
+	splitted_sgt->orig_nents = splitted_sgt->nents = i + 1;
+
+	pr_debug("%s:split[0x%px]reformatted, splitted:sgt->orig_nents=%d, nents=%d, out_nents[0]=%d\n",
+		__func__, par_dma_buf, splitted_sgt->orig_nents, splitted_sgt->nents, out_nents[0]);
+	/* parent's dupped sgt table is useless*/
+	dma_buf_detach(par_dma_buf, attach);
+	dma_buf_put(par_dma_buf);
+
+	return ret;
 }
 
-static const struct dma_buf_ops esw_common_buf_ops = {
-	.attach = esw_common_heap_attach,
-	.detach = esw_common_heap_detach,
-	.map_dma_buf = esw_common_map_dma_buf,
-	.unmap_dma_buf = esw_common_unmap_dma_buf,
-	.begin_cpu_access = esw_common_dma_buf_begin_cpu_access,
-	.end_cpu_access = esw_common_dma_buf_end_cpu_access,
-	.mmap = esw_common_mmap,
-	.vmap = esw_common_heap_vmap,
-	.vunmap = esw_common_heap_vunmap,
-	.release = esw_common_dma_buf_release,
+static void eswin_put_split_dmabuf(struct sg_table *splitted_sgt)
+{
+	kfree(splitted_sgt->sgl);
+}
+
+static void eswin_split_dma_buf_release(struct dma_buf *dmabuf)
+{
+	struct eswin_split_buffer *split_buffer = dmabuf->priv;
+	struct sg_table *table = &split_buffer->orig_sg_table;;
+
+	pr_debug("%s %d\n", __func__, __LINE__);
+	/* free splitted sgt->sgl which was allocated during esw_common_dmabuf_split_export()*/
+	eswin_put_split_dmabuf(table);
+
+	/* put parent's dmabuf which was got during esw_common_dmabuf_split_export()*/
+	dma_buf_put(split_buffer->dmabuf);
+
+	/* free split buffer which was allocated during esw_common_dmabuf_split_export()*/
+	kfree(split_buffer);
+}
+
+static const struct dma_buf_ops eswin_common_buf_ops = {
+	.attach = eswin_split_heap_attach,
+	.detach = eswin_split_heap_detach,
+	.map_dma_buf = eswin_split_map_dma_buf,
+	.unmap_dma_buf = eswin_spilt_unmap_dma_buf,
+	.begin_cpu_access = eswin_split_dma_buf_begin_cpu_access,
+	.end_cpu_access = eswin_split_dma_buf_end_cpu_access,
+	.mmap = eswin_split_mmap,
+	.vmap = eswin_split_heap_vmap,
+	.vunmap = eswin_split_heap_vunmap,
+	.release = eswin_split_dma_buf_release,
 };
 
-int esw_common_dmabuf_split_export(int dbuf_fd, unsigned int offset, size_t len, int fd_flags, char *name)
+
+int esw_common_dmabuf_split_export(int par_dmabuf_fd, unsigned int offset, size_t len, int fd_flags, char *name)
 {
-	struct esw_export_buffer_info *buffer_info;
+	struct eswin_split_buffer *split_buffer;
 	DEFINE_DMA_BUF_EXPORT_INFO(exp_info);
+	struct sg_table *table;
 	int fd;
 	struct dma_buf *dmabuf;
+	int ret = 0;
 
-	buffer_info = kzalloc(sizeof(*buffer_info), GFP_KERNEL);
-	if (!buffer_info)
+	split_buffer = kzalloc(sizeof(*split_buffer), GFP_KERNEL);
+	if (!split_buffer)
 		return -ENOMEM;
 
-	buffer_info->dbuf_fd = dbuf_fd;
-	buffer_info->fd_flags = fd_flags;
-	buffer_info->slice.offset = offset;
-	buffer_info->slice.len = len;
-	snprintf(buffer_info->name, sizeof(buffer_info->name), "%s", name);
+	INIT_LIST_HEAD(&split_buffer->attachments);
+	mutex_init(&split_buffer->lock);
+
+	split_buffer->dbuf_fd = par_dmabuf_fd;
+	split_buffer->fd_flags = fd_flags;
+	split_buffer->slice.offset = offset;
+	split_buffer->slice.len = len;
+	snprintf(split_buffer->name, sizeof(split_buffer->name), "%s", name);
 
-	buffer_info->dmabuf = dma_buf_get(buffer_info->dbuf_fd);
-	if (IS_ERR(buffer_info->dmabuf))
-		return PTR_ERR(buffer_info->dmabuf);
+	split_buffer->dmabuf = dma_buf_get(split_buffer->dbuf_fd);
+	if (IS_ERR(split_buffer->dmabuf)) {
+		kfree(split_buffer);
+		return PTR_ERR(split_buffer->dmabuf);
+	}
 
-//	printk("input slice: oft=0x%d, len=%lu\n", buffer_info->slice.offset, buffer_info->slice.len);
+	pr_debug("%s:%d, par_dbuf_fd %d, input slice: oft=0x%llx, len=0x%lx\n", __func__, __LINE__,
+		split_buffer->dbuf_fd, split_buffer->slice.offset, split_buffer->slice.len);
 
-	buffer_info->slice.offset = PAGE_ALIGN(buffer_info->slice.offset);
-	buffer_info->slice.len = PAGE_ALIGN(buffer_info->slice.len);
+	split_buffer->slice.offset = PAGE_ALIGN(split_buffer->slice.offset);
+	split_buffer->slice.len = PAGE_ALIGN(split_buffer->slice.len);
 
-//	printk("align slice: oft=0x%d, len=%lu\n", buffer_info->slice.offset, buffer_info->slice.len);
 
+	table = &split_buffer->sg_table;
+	ret = eswin_get_split_dmabuf(split_buffer);
+	if (ret) {
+		/* put back parent's dmabuf*/
+		dma_buf_put(split_buffer->dmabuf);
+		/* free split buffer */
+		kfree(split_buffer);
+		return ret;
+	}
 	/* create the dmabuf */
-	exp_info.exp_name = buffer_info->name;
-	exp_info.ops = &esw_common_buf_ops;
-	exp_info.size = buffer_info->slice.len;
-	exp_info.flags = buffer_info->fd_flags;
-	exp_info.priv = buffer_info;
+	exp_info.exp_name = split_buffer->name;
+	exp_info.ops = &eswin_common_buf_ops;
+	exp_info.size = split_buffer->slice.len;
+	exp_info.flags = O_RDWR | O_CLOEXEC;
+	exp_info.priv = split_buffer;
 
 	dmabuf = dma_buf_export(&exp_info);
 	if (IS_ERR(dmabuf)) {
+		eswin_put_split_dmabuf(table);
+		/* put back parent's dmabuf*/
+		dma_buf_put(split_buffer->dmabuf);
+		/* free split buffer */
+		kfree(split_buffer);
 		return PTR_ERR(dmabuf);
 	}
 
-	fd = dma_buf_fd(dmabuf, buffer_info->fd_flags);
+	fd = dma_buf_fd(dmabuf, split_buffer->fd_flags);
 	if (fd < 0) {
 		dma_buf_put(dmabuf);
-		/* put the splitted dmabuf, the esw_common_dma_buf_release will be called,
-		   the parent dmabuf will be put and the buffer_info will be free at that time */
+		/* put the splitted dmabuf, the eswin_split_dma_buf_release will be called,
+		   the parent dmabuf will be put and the split_buffer will be free at that time */
 	}
 	return fd;
 }
 
 EXPORT_SYMBOL(esw_common_dmabuf_split_export);
+
+/* create a misc device for attaching the parent dmabuf to get the parent's original sgt*/
+#define DRIVER_NAME 	"split_dmabuf"
+
+static int split_dmabuf_open(struct inode *inode, struct file *file)
+{
+
+	pr_debug("%s:%d, success!\n", __func__, __LINE__);
+
+	return 0;
+}
+
+static int split_dmabuf_release(struct inode *inode, struct file *file)
+{
+	pr_debug("%s:%d, success!\n", __func__, __LINE__);
+
+	return 0;
+}
+
+static struct file_operations split_dmabuf_fops = {
+	.owner        = THIS_MODULE,
+	.llseek        = no_llseek,
+	.open        = split_dmabuf_open,
+	.release    = split_dmabuf_release,
+};
+
+static struct miscdevice split_dmabuf_miscdev = {
+	.minor	= MISC_DYNAMIC_MINOR,
+	.name	= DRIVER_NAME,
+	.fops	= &split_dmabuf_fops,
+};
+
+static int __init split_dmabuf_init(void)
+{
+	int ret = 0;
+	struct device *dev;
+
+	ret = misc_register(&split_dmabuf_miscdev);
+	if(ret) {
+		pr_err ("cannot register miscdev (err=%d)\n", ret);
+		return ret;
+	}
+	split_dmabuf_dev = split_dmabuf_miscdev.this_device;
+	dev = split_dmabuf_dev;
+	if (!dev->dma_mask) {
+		dev->dma_mask = &dev->coherent_dma_mask;
+	}
+	ret = dma_set_mask_and_coherent(dev, DMA_BIT_MASK(48));
+	if (ret) {
+		pr_err("Unable to set coherent mask\n");
+		goto deregister_dev;
+	}
+
+	return 0;
+
+deregister_dev:
+	misc_deregister(&split_dmabuf_miscdev);
+	split_dmabuf_dev = NULL;
+	return ret;
+}
+
+static void __exit split_dmabuf_exit(void)
+{
+	misc_deregister(&split_dmabuf_miscdev);
+}
+
+module_init(split_dmabuf_init);
+module_exit(split_dmabuf_exit);
diff --git a/include/linux/dmabuf-heap-import-helper.h b/include/linux/dmabuf-heap-import-helper.h
index 0b0e5824f19f..9a474f6594e7 100644
--- a/include/linux/dmabuf-heap-import-helper.h
+++ b/include/linux/dmabuf-heap-import-helper.h
@@ -1,3 +1,24 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Header File of ESWIN DMABUF heap helper APIs
+ *
+ * Copyright 2024, Beijing ESWIN Computing Technology Co., Ltd.. All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Authors: Min Lin <linmin@eswincomputing.com>
+ */
+
 #ifndef _DMABUF_HEAP_IMPORT_H_
 #define _DMABUF_HEAP_IMPORT_H_
 
@@ -47,13 +68,20 @@ struct heap_mem {
 	size_t size;
 };
 
-struct esw_export_buffer_info {
-	char name[64];
-	int fd_flags;
-
-	int dbuf_fd;
-	struct dma_buf *dmabuf;
+struct eswin_split_buffer {
+	struct list_head attachments;
+	struct mutex lock;
+	unsigned long len;
+	struct sg_table sg_table; // for Re-formatted splitted sgt
+	struct sg_table orig_sg_table; // for originally splitted sgt
+	struct page **pages;
+	int vmap_cnt;
+	void *vaddr;
+	unsigned long fd_flags; // for vmap to determin the cache or non-cached mapping
 
+	char name[64]; // export name
+	int dbuf_fd; // parent dmabuf fd
+	struct dma_buf *dmabuf; // parent dmabuf
 	struct esw_slice_buffer {
 		__u64 offset;
 		size_t len;
-- 
2.48.1

